File: 71.简化路径.py

#
# @lc app=leetcode.cn id=71 lang=python3
#
# [71] 简化路径
#

# @lc code=start
class Solution:
    def simplifyPath(self, path: str) -> str:
        names=path.split('/') # str->list
        stack=list()

        for name in names:
            print(name)
            if name=='..':
                if stack:
                    stack.pop()
            elif name and name!='.':
                stack.append(name)
        return "/"+'/'.join(stack)

'''
class Solution:
    def simplifyPath(self, path: str) -> str:

        # 分割字符串为列表形式
        names = path.split("/")

        # 利用栈来处理
        stack = list()

        # 访问列表里面的元素
        for name in names:
            print(name)
            # 1、如果是 .. 
            if name == "..":
                # 在栈不为空的情况下
                if stack:
                    # 弹出栈顶元素
                    stack.pop()
            # 2、如果是有效值
            # 字母
            elif name and name != ".":
                stack.append(name)
        # stack = ['Python', 'World', 'Hello']
        # new_string = '/ '.join(stack)
        # print(new_string)
        # 输出结果 Python/ World/ Hello
        # 我们使用/ 作为分隔符，将栈stack中的元素连接成一个新的字符串。
        # 每个元素之间使用/ 进行分隔，因此输出结果中每个元素都以/ 结尾（除了最后一个元素）。
        return "/" + "/".join(stack)
'''
# @lc code=end

File: 71.简化路径.cpp
/*
 * @lc app=leetcode.cn id=71 lang=cpp
 *
 * [71] 简化路径
 */

// @lc code=start
class Solution {
public:
    string simplifyPath(string path) {
        std::vector<std::string> names;
        std::stringstream ss(path);
        std::string name;

        while (getline(ss,name,'/')){
            names.push_back(name);
        }

        std::vector<std::string> stack;

        for( const auto& name: names){
            // 常指针 &代表name可以改变names内容 auto自动推断name类型
            std::cout<<name<<std::endl;

            // if(name=='..'){
            if(name==".."){
                if (!stack.empty()){
                    stack.pop_back();
                }
            }
            // else if (!name.empty() && name!='.'){
            // ''和""不一样
            else if (!name.empty() && name!="."){
                stack.push_back(name);
            }
        }

        std::stringstream result;
        result<<"/";
        for (size_t i=0;i<stack.size();++i){
            result<<stack[i];
            if(i!=stack.size()-1){
                result<<'/';
            }
        }
        return result.str();
    }
};

/*
class Solution {
public:
    string simplifyPath(string path) {

        // 分割字符串为列表形式
        std::vector<std::string> names;
        std::stringstream ss(path);
        std::string name;
        
        while (getline(ss, name, '/')) {
            names.push_back(name);
        }

        // 利用栈来处理
        std::vector<std::string> stack;

        // 访问列表里面的元素
        for (const auto& name : names) {
            std::cout << name << std::endl;
            // 1、如果是 ..
            if (name == "..") {
                // 在栈不为空的情况下
                if (!stack.empty()) {
                    // 弹出栈顶元素
                    stack.pop_back();
                }
            }
            // 2、如果是有效值
            // 字母
            else if (!name.empty() && name != ".") {
                stack.push_back(name);
            }
        }

        std::stringstream result;
        result << "/";
        for (size_t i = 0; i < stack.size(); ++i) {
            result << stack[i];
            if (i != stack.size() - 1) {
                result << "/";
            }
        }

        return result.str();

    }
};
*/
// @lc code=end

